lib/ansible/modules/cloud/amazon/elasticsearch.py

```diff
@@ -0,0 +1,377 @@
+    if elasticserach_domain:
+        changed = False
+    else:
+        changed = True
+        _create_elasticsearch_domain(connection, module)
+        elasticserach_domain = is_present(connection, domain)
+    return changed, elasticserach_domain
+
+
+def is_present(connection, name):
+    """Get the current status of Elasticsearch domain"""
+    try:
+        response = connection.describe_elasticsearch_domain(
+            DomainName=name
+        )
+
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == 'ResourceNotFoundException':
+            return None
+        else:
+            raise
+
+    return response
+
+
+def check_existing_config(module, domain):
+    modifications_needed = collections.defaultdict(dict)
+
+    existing_config = domain['DomainStatus']
+    desired_config = module.params
+
+    config_to_check = [
+        (('ElasticsearchVersion',), desired_config['version']),
+        (('ElasticsearchClusterConfig', 'InstanceCount'),
+            desired_config['instance_count']),
+        (('ElasticsearchClusterConfig', 'InstanceType'),
+            desired_config['instance_type']),
+        (('ElasticsearchClusterConfig', 'ZoneAwarenessEnabled'),
+            desired_config['zone_awareness']),
+        (('ElasticsearchClusterConfig', 'DedicatedMasterEnabled'),
+            desired_config['dedicated_master_enabled']),
+        (('EBSOptions', 'VolumeType'), desired_config['ebs_volume_type']),
+        (('EBSOptions', 'VolumeSize'), desired_config['ebs_volume_size']),
+    ]
+
+    if desired_config['dedicated_master_enabled']:
+        config_to_check.extend(
+            [
+                (('ElasticsearchClusterConfig', 'DedicatedMasterType'),
+                    desired_config['dedicated_master_type']),
+                (('ElasticsearchClusterConfig', 'DedicatedMasterCount'),
+                    desired_config['dedicated_master_count']),
+            ]
+        )
+
+    try:
+        for key_path, desired_value in config_to_check:
+            # Traverse key path to get existing value
+            existing_value = existing_config
+            for key in key_path:
+                existing_value = existing_value.get(key)
+
+            if existing_value != desired_value:
+                # Traverse key path in modifications dictionary to submit to AWS
+                modification = modifications_needed
+                for key in key_path[:-1]:
+                    if key not in modification:
+                        modification[key] = {}
+
+                    modification = modification[key]
+
+                modification[key_path[-1]] = desired_value
+
+    except Exception as e:
+        module.fail_json(
+            msg=str(e), key_path=key_path, desired_value=desired_value,
+            existing_config=existing_config,
+            modifications_needed=modifications_needed)
+
+    # Config is JSON, so compare JSON dictionaries instead of strings
```

Comment:
You can compare two dicts with compare_policies() since it fixes comparisons with lists of length one vs a string and different dictionary order:
`from ansible.module_utils.ec2 import compare_policies`
```
if compare_policies(existing_access_json, supplied_access_json):
    modifications_needed['AccessPolicies'] = module.params['access_policies']
```